SIL.Media/Naudio/AudioRecorder.cs

```diff
@@ -279,44 +326,82 @@ public virtual void BeginRecording(string waveFileName)
+				// monitoring and it happens to get there first, the other will find it is no
+				// longer in a valid state to begin monitoring but will happily find that it
+				// is now in a valid state to begin recording. If the one that is trying to
+				// begin recording gets there first, the other one will get the exception.
+				// In any case, handling an InvalidOperationException is the caller's
+				// responsibility.
+				BeginMonitoringIfNeeded();
 
-				if (_recordingState != RecordingState.Monitoring)
+				lock (this)
 				{
-					throw new InvalidOperationException("Can't begin recording while we are in this state: " + _recordingState.ToString());
-				}
+					if (_recordingState != RecordingState.Monitoring)
+						throw new InvalidRecordingStateException();
+
+					// Kind of a weird race condition here, but for some reason we allow the
+					// caller to change the buffers without getting a lock or checking the state.
+					// (Even more oddly, if this happens while recording, it doesn't take effect
+					// until the next time they start recording, but if it happens in the
+					// nanoseconds between calling BeginMonitoring_Internal and here, then we
+					// reset. But this was the existing behavior, and changing it now could break
+					// some exiting code.)
+					if (_waveInBuffersChanged)
+					{
+						CloseWaveIn();
+						if (!BeginMonitoringIfNeeded())
+							throw new InvalidRecordingStateException();
+					}
 
-				if (_waveInBuffersChanged)
-				{
-					CloseWaveIn();
-					BeginMonitoring();
-				}
+					_bytesRecorded = 0;
+
+					WaveFileWriter writer;
+					if (!File.Exists(waveFileName) || !appendToFile)
+						writer = new WaveFileWriter(waveFileName, _recordingFormat);
+					else
+					{
+						var buffer = GetAudioBufferToAppendTo(waveFileName);
+						writer = new WaveFileWriter(waveFileName, _recordingFormat);
+						writer.Write(buffer, 0, buffer.Length);
+					}
 
-				_bytesRecorded = 0;
+					_fileWriterThread = new FileWriterThread(writer);
 
-				WaveFileWriter writer;
-				if (!File.Exists(waveFileName) || !appendToFile)
-					writer = new WaveFileWriter(waveFileName, _recordingFormat);
-				else
-				{
-					var buffer = GetAudioBufferToAppendTo(waveFileName);
-					writer = new WaveFileWriter(waveFileName, _recordingFormat);
-					writer.Write(buffer, 0, buffer.Length);
+					_recordingStartTime = DateTime.Now;
+					_prevRecordedTime = 0d;
+					RecordingState = RecordingState.Recording;
 				}
-				_fileWriterThread = new FileWriterThread(writer);
 
-				_recordingStartTime = DateTime.Now;
-				_prevRecordedTime = 0d;
-				RecordingState = RecordingState.Recording;
+				if (RecordingStarted != null)
+					RecordingStarted(this, EventArgs.Empty);
+			}
+			catch (InvalidRecordingStateException e)
+			{
+				var error = new InvalidOperationException(
+					$"Can't begin recording while we are in this state: {_recordingState}", e);
```

Comment:
Technically it would be better to capture _recordingState in a local variable inside the lock (or as part of the InvalidRecordingStateException). A race condition could conceivably change _recordingState between when you threw the InvalidRecordingStateException and when you now try to report the reason for it, resulting in a nonsensical message.
It would also be OK for this first catch to be inside the lock. It's the NotifyUserOfProblem that wants to be outside it.